Detectors/TPC/reconstruction/src/HwClusterer.cxx

```diff
@@ -8,461 +8,530 @@
+
+  ClusterHardware cluster;
+  std::vector<std::pair<MCCompLabel, unsigned>> mcLabels;
+
+  unsigned qMaxIndex;
+  for (unsigned short row = mNumRows; row--;) {
+    // two empty pads on the left and right without a cluster peak
+    for (short pad = 2; pad < mPadsPerRow[row] - 2; ++pad) {
+      qMaxIndex = (timebin % 5) * mPadsPerRow[row] + pad;
+      // Check if pad is above peak threshold already here to avoid function
+      // call in most of the case. Provide index of center to avoid
+      // recalculation of the same index.
+      if (mDataBuffer[row][qMaxIndex] > (mPeakChargeThreshold << 4) && hwClusterFinder(qMaxIndex, pad, timebin, row, cluster, mcLabels)) {
+        mTmpClusterArray[mGlobalRowToRegion[row]]->emplace_back(cluster, std::move(mcLabels));
+        mcLabels.clear();
+      }
+    }
+  }
+}
 
-   /*
-   * vector to store threads for parallel processing
-   */
-  std::vector<std::thread> thread_vector;
-
-  LOG(DEBUG) << "Starting " << mNumThreads << " threads, hardware supports " << std::thread::hardware_concurrency() << " parallel threads." << FairLogger::endl;
-
-  for (unsigned threadId = 0; threadId < std::min(mNumThreads, mCRUMax + 1); ++threadId) {
-    struct CfConfig cfConfig = {
-      threadId,
-      mNumThreads,
-      mCRUMin,
-      mCRUMax,
-      static_cast<unsigned>(mPadsMax)+2+2,
-      iTimeBinMin,
-      iTimeBinMax,
-      mEnableNoiseSim,
-      mEnablePedestalSubtraction,
-      mIsContinuousReadout,
-      mNoiseObject,
-      mPedestalObject
-    };
-    thread_vector.emplace_back(
-        processDigits,                          // function name
-        std::ref(mDigitContainer),              // digit container for individual CRUs
-        std::ref(mClusterFinder),               // cluster finder for individual CRUs
-        std::ref(mClusterStorage),              // container to store found clusters
-        std::ref(mClusterDigitIndexStorage),    // container to store found cluster MC Labels
-        cfConfig                                // configuration
-        );
+//______________________________________________________________________________
+void HwClusterer::finishFrame(bool clear)
+{
+  unsigned HB;
+  // Search in last remaining timebins for clusters
+  for (int i = mLastTimebin; i - mLastTimebin < 5; ++i) {
+    HB = i < 2 ? 0 : (i - 2) / 447; // integer division on purpose
+    if (HB != mLastHB) {
+      writeOutputWithTimeOffset(mLastHB * 447);
+    }
+
+    findClusterForTime(i - 2);
+    clearBuffer(i + 1);
+    mLastHB = HB;
   }
+  writeOutputWithTimeOffset(mLastHB * 447);
 
+  if (clear) {
+    for (auto i : { 0, 1, 2, 3, 4 })
+      clearBuffer(i);
+  }
+}
 
-  /*
-   * wait for threads to join
-   */
-  for (std::thread& t: thread_vector) {
-    t.join();
+//______________________________________________________________________________
+void HwClusterer::clearBuffer(int timebin)
+{
+  mMCtruth[timebin % 5] = std::shared_ptr<MCLabelContainer const>(nullptr);
```

Comment:
Wouldn't a simple `.reset();` be more appropriate? Overwriting with a `nullptr` looks a bit awkward